### PR TITLE
add metric capture to js

### PR DESCRIPTION
**Pull request by imsickofmaps**
## SMS
- [x] total unique users (text widget)
- [x] total registered users (text widget)
- [x] total unique users and registered users per week (line graph widget)
- [x] total girl registered users
- [x] total messages sent (text widget)
- [x] total messages delivered (text widget) (will have to be Vumi Go)
- [x] total messages received (text widget)
## USSD
- [x] total unique users
- [x] total registrations (text widget)
- [x] total users and total registrations per week (line graph widget)
- [x] total girl registered users
- [x] total sessions initiated (text widget) (will have to be Vumi Go)
- [x] total sessions per week and total unique users who initiated a session per week (line graph widget) (will have to be Vumi Go)
- [x] articles views per week (line graph widget)
- [x] shangazi advice views per week (line graph widget)
- [x] directory views per week (line graph widget)
- [x] opinions views per week (line graph widget)
- [x] quiz views per week (line graph widget)

### CODE REVIEW

**Comment by imsickofmaps** [comment]
@smn or @hodgestar - ready for review 

cc: @sheetalgordhan

File: js_sandbox/lib/go-nike-ghr-sms.js [hunk 1]
```diff
@@ -161,6 +177,73 @@ function GoNikeGHRSMS() {
...
+
+    self.get_week_commencing = function(today) {
+        // today should be var today = new Date();
+        // Creates a value like "2013-06-10"
+        var date = self.get_monday(today);
+        return date.toISOString().substring(0,10);
+    };
+
+
+    self.increment_counter = function(metric_key) {
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr-sms.js]: Is this function being called anywhere?

File: js_sandbox/lib/go-nike-ghr-sms.js [hunk 2]
```diff
@@ -161,6 +177,73 @@ function GoNikeGHRSMS() {
...
+                agg: 'max'
+            });
+        });
+        return p;
+    };
+
+
+
+    self.check_u18_girl = function(){
+        var p = self.crm_get('userinteraction/?msisdn='+im.user_addr+'&feature=REGISTRATION&format=json');
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr-sms.js]: Why not use your `self.url_encode()` function here?

``` javascript
var p = self.crm_get('userinteraction/?' + self.url_encode({msisdn: im.user_addr, feature: 'REGISTRATION', format: 'json'}));
```

File: js_sandbox/lib/go-nike-ghr.js [hunk 3]
```diff
@@ -72,6 +72,37 @@ function GoNikeGHR() {
         );
     };
 
+    self.increment_counter = function(metric_key) {
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr.js]: Doesn't look like it's being used.

File: js_sandbox/lib/go-nike-ghr.js [hunk 4]
```diff
@@ -476,7 +521,41 @@ function GoNikeGHR() {
...
+                    var contact_key;
+                    var update_req = false;
+                    
+                    var p_c = self.get_contact(im);
+                    p_c.add_callback(function(result){
+                        contact_key = result.contact.key;
+                        if (result.contact["extras-ghr_last_active_week"] !== undefined){
+                            if (new Date(wc) > new Date(result.contact["extras-ghr_last_active_week"])){
+                                update_req = true;
+                                return self.increment_and_fire_direct("ghr_ussd_total_users_"+wc);
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr.js]: I think you're potentially going to run into trouble with this `update_req` variable. I think using a `Promise` here would be a better approach:

``` javascript
var piafd = self.increment_and_fire_direct("ghr_ussd_total_users_"+wc);
piafd.add_callback(function(result)) {
    return true;
}
return piafd;
```

File: js_sandbox/lib/go-nike-ghr.js [hunk 5]
```diff
@@ -476,7 +521,41 @@ function GoNikeGHR() {
...
+                            }
+                        } else { // for contacts somehow missing attribute
+                            update_req = true;
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr.js]: here I'd just `return true;`

File: js_sandbox/lib/go-nike-ghr.js [hunk 6]
```diff
@@ -476,7 +521,41 @@ function GoNikeGHR() {
...
+                        }
+                    });
+                    p_c.add_callback(function(result){
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr.js:540]: `result` here would be the value of `update_req` which makes this code a little bit more predictable.

File: js_sandbox/lib/go-nike-ghr.js [hunk 7]
```diff
@@ -553,7 +634,17 @@ function GoNikeGHR() {
...
-                        ]
+                        ],
+                        null,
+                        {
+                            on_enter: function() {
+                                // Metric counting and logging
+                                var p_c = new Promise();
+                                p_c.add_callback(self.increment_and_fire("ghr_ussd_total_unique_users"));
+                                p_c.callback();
+                                return p_c;
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr.js]: Why not return the Promise that `increment_and_fire_direct()` returns here?

File: js_sandbox/lib/go-nike-ghr.js [hunk 8]
```diff
@@ -678,15 +770,17 @@ function GoNikeGHR() {
...
-                                var p_log = self.interaction_log("REGISTRATION", "gender", gender);
-                                p_log.add_callback(function() {
-                                    var p_log2 = self.interaction_log("REGISTRATION", "age", age);
-                                    p_log2.add_callback(function() {
-                                        var p_log3 = self.interaction_log("REGISTRATION", "sector", sector);
-                                        return p_log3;
-                                    });
-                                    return p_log2;
+                                var p_log = new Promise();
+                                p_log.add_callback(function(){self.interaction_log("REGISTRATION", "gender", gender);});
```
Comment (smn) [inline js_sandbox/lib/go-nike-ghr.js]: You want to return the `Promise` that `self.interaction_log()` returns here (and all the other places you're doing this).

**Comment by smn** [comment]
Looks good in general, just some comments on `Promise`s not being returned and I think a function isn't being used.

**Comment by imsickofmaps** [comment]
@smn ready for re-review unless you're happy that I've made the changes you requested

**Comment by smn** [comment]
:+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake: :+1: :cake:

**Pull request merged**